Feature/sort ifs shortcuts

chrjorgensen opened the pull request:
### Changes

This PR will add the ability to sort the IFS shortcuts in the IFS browser:

- The shortcuts have a new item type `shortcut` - previously they were `directory` like regular directories.
- There are new icons for the shortcuts to move the shortcut up or down in the list, to the top or bottom and to sort the list.
- There is a new icon in the view to sort the shortcuts alphabetically - case-insensitive.
- There is a new setting in the configuration to automatically sort the shortcuts when a shortcut is added or removed.

It was my ambition to have a drag-and-drop function implemented so one could just reorder by dragging the shortcut into he desired place - but this is currently way above my skills in VS Code extension programming! 😆 

I use the IFS browser almost all the time - with my sources in IFS - so that's why I have focused my contributions to this area of your excellent VS Code extension. My many IFS shortcuts had been added in a long time-frame and were in a chaotic order, and I added this functionality for my own convenience, but it could be useful to others as well. I know I could have manually ordered the shortcut list in the configuration, but this may not be obvious to all...

### Checklist

* [x] have tested my change
* [ ] updated relevant documentation
* [x] Remove any/all `console.log`s I added
* [x] eslint is not complaining
* [ ] have added myself to the contributors' list in [CONTRIBUTING.md](https://github.com/halcyon-tech/vscode-ibmi/blob/master/CONTRIBUTING.md)
* [x] **for feature PRs**: PR only includes one feature enhancement.

worksofliam commented:
Wow, this is some excellent work. Expect a review over the weekend!

> It was my ambition to have a drag-and-drop function implemented so one could just reorder by dragging the shortcut into he desired place - but this is currently way above my skills in VS Code extension programming! 

I would love this also. This is actually not possible with the VS Code API at the moment... so it's not possible no matter how bad we want that!

worksofliam commented on src/views/ifsBrowser.js:
```diff
@@ -110,14 +112,155 @@ module.exports = class ifsBrowserProvider {
+      vscode.commands.registerCommand(`code-for-ibmi.moveIFSShortcutDown`, async (node) => {
+        const config = instance.getConfig();
+
+        let moveDir;
```
I think this can be re-written as 

```
if (node) {
  const moveDir = node.path ? node.path.trim() || null;
  
  if (moveDir) {
    // logic
  }
}

worksofliam replied:
I see you've followed this format in a few commands, too.

chrjorgensen replied:
Fixed in commit [54f0421](https://github.com/halcyon-tech/vscode-ibmi/pull/708/commits/54f0421c703896eba43012c234627af261ded7f9)

worksofliam commented on src/api/Configuration.js:
```diff
@@ -26,6 +26,9 @@ module.exports = class Configuration {
     this.ifsShortcuts = base.ifsShortcuts || [];
 
+    /** @type {boolean} Undefined means not created, so default to on */
+    this.autoSortIFSShortcuts = (base.autoSortIFSShortcuts === true );
```
I see you've added the config here, which is correct. But, you've also got to add it to the `package.json`. 

It belongs in `configuration` -> `code-for-ibmi.connectionSettings`. Look for `autoConvertIFSccsid` as an example.

chrjorgensen replied:
Fixed in commit [78bd326](https://github.com/halcyon-tech/vscode-ibmi/pull/708/commits/78bd326a9dab8e768172cba5573475cf21254507)

worksofliam reviewed (changes requested):
It's very good!! Seriously, a great PR. Please have a review of my comments.

Also, for the right click options, I don't think 

* move up
* move down
* sort all

belong as right-click options on the `shortcut`. This is because they're already accessible from the UI and I feel as if it adds some clutter. I do like the move to top and bottom, though.

<img width="490" alt="image" src="https://user-images.githubusercontent.com/3708366/173192821-60e27d1a-2a2b-4c3a-94b1-e9a4c4466904.png">

chrjorgensen commented:
Thanks for the feedback - and fair comments.
I'll have a look at it asap - hopefully this weekend...

chrjorgensen commented:
I agree the right-click options for moving a shortcut up and down and sorting the shortcuts may be redundant to the UI, and I have removed them now. I added them because the icons for moving a shortcut up and down stays on the display if the shortcut is clicked, and I don't know if I like that... 😄

chrjorgensen commented:
About the drag-and-drop - are you sure it is not possible with the current VS Code API?

It is my understanding, that it is possible (see this [example](https://github.com/microsoft/vscode-extension-samples/blob/799d4ab2499ff4d70361e77a1bfde4388114a79b/tree-view-sample/src/testViewDragAndDrop.ts)) but it requires the use of `createTreeView` and coding of a `dragAndDropController` instead of the simple `registerTreeDataProvider` - and that is a change way too big for me to code...!

worksofliam commented:
@chrjorgensen Interesting. As far as I'm aware it wasn't and I think it would require implementing our own provider. I will ask some of the other extension authors I know and see what they think.

worksofliam commented:
@chrjorgensen Your changes look excellent! Thanks. If you are happy with everything, let me know and we can get this merged.

chrjorgensen commented:
I'm happy - go ahead... 👍
